### ISSUE TITLE
Replacing lists via node attributes

### ISSUE THREAD

**Issue by tbsuht** [CESNET/libyang#2219]
I was not sure about creating this at the sysrepo or libyang repo, because I use both APIs in my example, so I will try here.

I want to replace two lists (and only the lists) in two different models with a single operation.
I found this helpful issue which shows how to replace specific data nodes: https://github.com/sysrepo/sysrepo/issues/1809

I'm using this model for testing:
```
module basic_test_list {
    namespace "urn:basic_test_list";
    prefix e;
    container foo {
        list bar {
            key k;
            leaf k {
                type string;
            }
            leaf val1 {
                type string;
            }
            leaf val2 {
                type string;
            }
        }
    }
}
```

I have replicated this as `basic_test_list` and `basic_test_list2` in order to have two models to work with.
Now I'm filling the datastore with whatever data.

And then I'm trying to do this to replace the complete lists, but somehow there are list entries added and the list is not replaced.
```
{
   "basic_test_list:foo":{
      "bar":[
         {
            "k":"new"
         }
      ],
      "@bar":{
         "ietf-netconf:operation":"replace"
      }
   },
   "basic_test_list2:foo":{
      "bar":[
         {
            "k":"new2"
         }
      ],
      "@bar":{
         "ietf-netconf:operation":"replace"
      }
   }
}
```

Basically the following functions are called in that order:
```
lyd_parse_data()
sr_edit_batch(default_operation='merge')
apply_changes()
```

Is there something obviously wrong with what I'm doing?

libyang/2.1.148
sysrepo/2.2.150

**Comment by michalvasko** [comment]
Try using the [correct syntax](https://datatracker.ietf.org/doc/html/rfc7952#section-5.2.2) for JSON metadata of lists, it should work then.

**Comment by tbsuht** [comment]
Hm, does it look wrong? In order to replace a list, the metadata should be a sibling of the list, which it is.

**Comment by michalvasko** [comment]
That is not what the RFC said, you even have an example there.

>  For a data node instance that is encoded as a JSON object (i.e., a
   container, list entry, or anydata node), the metadata object is added
   as a new member of that object with the name "@".

**Comment by tbsuht** [comment]
Sorry, I don't get it, because it says that the data node is a JSON object, but I need a JSON array.
I can't add an attribute directly to an array, that's why I thought it must be a sibling of the array.

I tried now to add it as a sibling, as an attribute (which can't be parsed), and as attribute of a list element (which I don't want, because the whole list should be replaced.).

I also found this test case which is also passing it as a sibling?
https://github.com/CESNET/libyang/blob/4c733412e7173219166be7053940326a92699765/tests/utests/data/test_parser_json.c#L678

**Comment by jktjkt** [comment]
I think that you're hitting the same issue which basically [prevents deleting an entire list through NETCONF or RESTCONF](https://github.com/netconf-wg/netconf-next/issues/35). As far as I can tell, in YANG, there's no concept of "a `list`" or "a `leaf-list`"; there are only "`list` entries" and "`leaf-list` entries" (sometimes they are called "`list` instances" or "`leaf-list` instances" to create even more confusion). See the [terminology from the YANG 1.1 RFC](https://datatracker.ietf.org/doc/html/rfc7950#section-3):

> - `leaf`: A data node that exists in at most one instance in the data tree.  A leaf has a value but no child nodes.
>
> - `leaf-list`: Like the leaf node but defines a set of uniquely identifiable nodes rather than a single node.  Each node has a value but no child nodes.
>
> -  `list`: An interior data node that may exist in multiple instances in the data tree.  A list has no value, but rather a set of child  nodes.

 Then everything is kind-of-obscured by the JSON encoding which makes one think that there actually *is* a special node for an entire `list`/`leaf-list`, but nope. It's just a set of individual nodes, one per each list entry, and in the abstract tree of YANG data nodes, there **is not** a special item for a `list` (or a `leaf-list`) that you can use to manipulate all of the **entries** of that `list`/`leaf-list`.

That means that your "replace an entire `list`" operation will have to create one `list` entry for each list instance (containing at least all the keys; the actual values of other data nodes in the `list` instance are not meaningful), and tag each of these `list` entries with the `delete`/`remove` metadata.

(As far as I understand the spec. If I'm wrong, it won't be the first time.)

**Comment by michalvasko** [comment]
> I also found this test case which is also passing it as a sibling?

Because in that test `z` is a **leaf**, not a **list**.

> which I don't want, because the whole list should be replaced.

As @jktjkt said, this is not possible based on the NETCONF specs. However, *sysrepo* support some "extensions" and defines additional operations that can be used via NETCONF as well. In this case you can use `"sysrepo:operation":"purge"` for any single list instance, which should cause all the list instances to be removed. You should then be able to create specific new instances, perhaps even in the same `edit-config`, I am not completely sure about that. Just note that this is not going to work at all for other NETCONF servers and is specific to ours.

**Comment by tbsuht** [comment]
> I think that you're hitting the same issue which basically [prevents deleting an entire list through NETCONF or RESTCONF](https://github.com/netconf-wg/netconf-next/issues/35). As far as I can tell, in YANG, there's no concept of "a `list`" or "a `leaf-list`"; there are only "`list` entries" and "`leaf-list` entries" (sometimes they are called "`list` instances" or "`leaf-list` instances" to create even more confusion). See the [terminology from the YANG 1.1 RFC](https://datatracker.ietf.org/doc/html/rfc7950#section-3):
> 
> > * `leaf`: A data node that exists in at most one instance in the data tree.  A leaf has a value but no child nodes.
> > * `leaf-list`: Like the leaf node but defines a set of uniquely identifiable nodes rather than a single node.  Each node has a value but no child nodes.
> > * `list`: An interior data node that may exist in multiple instances in the data tree.  A list has no value, but rather a set of child  nodes.
> 
> Then everything is kind-of-obscured by the JSON encoding which makes one think that there actually _is_ a special node for an entire `list`/`leaf-list`, but nope. It's just a set of individual nodes, one per each list entry, and in the abstract tree of YANG data nodes, there **is not** a special item for a `list` (or a `leaf-list`) that you can use to manipulate all of the **entries** of that `list`/`leaf-list`.
> 
> That means that your "replace an entire `list`" operation will have to create one `list` entry for each list instance (containing at least all the keys; the actual values of other data nodes in the `list` instance are not meaningful), and tag each of these `list` entries with the `delete`/`remove` metadata.
> 
> (As far as I understand the spec. If I'm wrong, it won't be the first time.)

Thanks for the explanation!

> > I also found this test case which is also passing it as a sibling?
> 
> Because in that test `z` is a **leaf**, not a **list**.
> 
> > which I don't want, because the whole list should be replaced.
> 
> As @jktjkt said, this is not possible based on the NETCONF specs. However, _sysrepo_ support some "extensions" and defines additional operations that can be used via NETCONF as well. In this case you can use `"sysrepo:operation":"purge"` for any single list instance, which should cause all the list instances to be removed. You should then be able to create specific new instances, perhaps even in the same `edit-config`, I am not completely sure about that. Just note that this is not going to work at all for other NETCONF servers and is specific to ours.

Cool! This is working. 

For sake of completeness the following is working:

```
{
   "basic_test_list:foo":{
      "bar":[
         {
            "@":{
               "sysrepo:operation":"purge"
            },
            "k":"foobar"
         },
         {
            "k":"new"
         }
      ]
   },
   "basic_test_list2:foo":{
      "bar":[
         {
            "@":{
               "sysrepo:operation":"purge"
            },
            "k":"foobar2"
         },
         {
            "k":"new2"
         }
      ]
   }
}
```

Thanks @jktjkt and @michalvasko !
